Rename static_pointer_cast() / static_reference_cast() to unsafeRefPtrDowncast() / unsafeRefDowncast()

cdumez opened the pull request:
#### ee6d739d218f27a98346d64c683e4b1bcd3c6b84
<pre>
Rename static_pointer_cast() / static_reference_cast() to unsafeRefPtrDowncast() / unsafeRefDowncast()
<a href="https://bugs.webkit.org/show_bug.cgi?id=301991">https://bugs.webkit.org/show_bug.cgi?id=301991</a>

Reviewed by Darin Adler.

Rename static_pointer_cast() / static_reference_cast() to unsafeRefPtrDowncast()
/ unsafeRefDowncast() to make it clear they are unsafe and reduce their use in
the codebase by leveraging:
1. downcast&lt;&gt;() which is truly safe and supports smart pointers nowadays
2. uncheckedDowncast&lt;&gt;() which is not ideal but better since it will at least validate
   the cast in debug builds.
3. New upcast() functions which are safe since it is safe to use static_cast for upcasts.

* Source/JavaScriptCore/bytecode/MetadataTable.h:
(JSC::MetadataTable::unlinkedMetadata const):
* Source/JavaScriptCore/inspector/InspectorProtocolTypes.h:
(Inspector::Protocol::BindingTraits&lt;JSON::ArrayOf&lt;T&gt;&gt;::runtimeCast):
* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py:
* Source/JavaScriptCore/inspector/scripts/tests/expected/type-requiring-runtime-casts.json-result:
* Source/WTF/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WTF/wtf/Ref.h:
(WTF::upcast):
(WTF::unsafeRefDowncast):
(WTF::uncheckedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
(WTF::static_reference_cast): Deleted.
* Source/WTF/wtf/RefPtr.h:
(WTF::upcast):
(WTF::unsafeRefPtrDowncast):
(WTF::uncheckedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
(WTF::static_pointer_cast): Deleted.
* Source/WTF/wtf/WeakPtr.h:
(WTF::weak_ptr_impl_cast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/WeakRef.h:
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/text/AtomString.h:
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::AtomStringImpl::addSlowCase):
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::AtomStringImpl::add):
* Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebCore/bindings/js/JSDOMWrapperCache.h:
(WebCore::createWrapper):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::convertToExceptionOrNumericValue):
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementFromPoint):
(WebCore::TreeScope::elementsFromPoint):
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::splitAncestorsWithUnicodeBidi):
* Source/WebCore/editing/DeleteSelectionCommand.cpp:
(WebCore::firstInSpecialElement):
(WebCore::lastInSpecialElement):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setComposition):
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplaceSelectionCommand::handleStyleSpans):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::lastTypingCommandIfStillOpenForTyping):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::finalizeFrame):
(WebCore::InspectorCanvas::markCurrentFrameIncomplete):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateDragAndDrop):
* Source/WebCore/page/ImageOverlayController.cpp:
(WebCore::ImageOverlayController::selectionQuadsDidChange):
* Source/WebCore/page/scrolling/ThreadedScrollingCoordinator.cpp:
(WebCore::ThreadedScrollingCoordinator::pageDestroyed):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
* Source/WebCore/rendering/RenderPtr.h:
(WTF::downcast):
(WebCore::static_pointer_cast): Deleted.
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
(WebCore::StyleCrossfadeImage::computedStyleValue const):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::computedStyleValue const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):
* Source/WebCore/svg/properties/SVGPrimitivePropertyAnimator.h:
(WebCore::SVGPrimitivePropertyAnimator::SVGPrimitivePropertyAnimator):
* Source/WebCore/svg/properties/SVGValuePropertyAnimator.h:
(WebCore::SVGValuePropertyAnimator::SVGValuePropertyAnimator):
* Source/WebCore/svg/properties/SVGValuePropertyListAnimator.h:
(WebCore::SVGValuePropertyListAnimator::SVGValuePropertyListAnimator):
* Source/WebKit/Shared/API/APIArray.h:
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
(webkitFeatureCreate):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
(webkitWebExtensionMatchPatternCreate):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::experimentalFeatures):
(WebKit::WebPreferences::internalDebugFeatures):
* Tools/TestWebKitAPI/Tests/WTF/Ref.cpp:
(TestWebKitAPI::TEST(WTF_Ref, StaticReferenceCastFromConstReference)):
(TestWebKitAPI::TEST(WTF_Ref, StaticReferenceCastFromRValueReference)):

Canonical link: <a href="https://commits.webkit.org/302695@main">https://commits.webkit.org/302695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c268e2513feb55c7691f924646a08fad376cff0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137146 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81226 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/16ef8398-39cb-496a-bb69-09978b463296) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98853 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66680 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/08b4b120-c8aa-4902-88b8-cfc2b1d22192) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1501 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116216 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79533 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c5077311-70f2-4d4e-aa26-717889c2518b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1418 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34349 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80419 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121749 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109899 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139629 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128209 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1694 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107359 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112565 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107234 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27333 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1471 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31055 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54566 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1885 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65254 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161223 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1699 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40201 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1734 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1806 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

iidmsa commented on Source/WTF/wtf/Ref.h:
```diff
@@ -268,15 +268,35 @@ inline Ref<X, _PtrTraits, RefDerefTraits> Ref<X, _PtrTraits, RefDerefTraits>::re
-inline Ref<X, _PtrTraits, RefDerefTraits> static_reference_cast(Ref<Y, _OtherPtrTraits, OtherRefDerefTraits>&& reference)
+inline Ref<X, _PtrTraits, RefDerefTraits> static_reference_upcast(Ref<Y, _OtherPtrTraits, OtherRefDerefTraits>&& reference)
 {
+    static_assert(!std::is_same_v<Y, X>, "Unnecessary cast to same type");
```
`std::same_as`?

iidmsa commented on Source/WTF/wtf/Ref.h:
```diff
@@ -268,15 +268,35 @@ inline Ref<X, _PtrTraits, RefDerefTraits> Ref<X, _PtrTraits, RefDerefTraits>::re
+inline Ref<X, _PtrTraits, RefDerefTraits> static_reference_upcast(Ref<Y, _OtherPtrTraits, OtherRefDerefTraits>&& reference)
 {
+    static_assert(!std::is_same_v<Y, X>, "Unnecessary cast to same type");
+    static_assert(std::is_base_of_v<X, Y>, "Should be an upcast");
```
`std::derived_from`?

darinadler reviewed (approved):
My only complaint is there are too many words in these names. We don't need both _unsafe_ and _static_ in the name of a cast. Also seems unneeded to have the word _static_ and the word _upcast_. Also, as we get farther from the semantics of casts built into the C++ library I think we can use `interCaps` style.

darinadler commented on Source/WTF/wtf/Ref.h:
```diff
@@ -268,15 +268,35 @@ inline Ref<X, _PtrTraits, RefDerefTraits> Ref<X, _PtrTraits, RefDerefTraits>::re
-inline Ref<X, _PtrTraits, RefDerefTraits> static_reference_cast(Ref<Y, _OtherPtrTraits, OtherRefDerefTraits>&& reference)
+inline Ref<X, _PtrTraits, RefDerefTraits> upcast(Ref<Y, _OtherPtrTraits, OtherRefDerefTraits>&& reference)
 {
+    static_assert(!std::same_as<Y, X>, "Unnecessary cast to same type");
+    static_assert(std::derived_from<Y, X>, "Should be an upcast");
```
Are these better as assertions or requirements? I guess maybe the error messages are easier to read when they are assertions. But requirements might allow us to use more overloading and fewer different function names.

darinadler commented on Source/WTF/wtf/text/AtomStringImpl.h:
```diff
@@ -130,9 +130,9 @@ ALWAYS_INLINE Ref<AtomStringImpl> AtomStringImpl::add(StringImpl& string)
 ALWAYS_INLINE Ref<AtomStringImpl> AtomStringImpl::add(Ref<StringImpl>&& string)
 {
-    if (string->isAtom()) {
+    if (is<AtomStringImpl>(string)) {
```
Why not `dynamicDowncast`?

cdumez replied:
There is not a way to write this code as efficiently with dynamicDowncast or without introducing a use-after-move:
```
if (RefPtr atomStringImpl = dynamicDowncast<AtomStringImpl>(string)) {
    return atomStringImpl;
return addSlowCase(WTFMove(string));
```
Slower because it causes an extra ref of the StringImpl.

```
if (RefPtr atomStringImpl = dynamicDowncast<AtomStringImpl>(WTFMove(string))) {
    return atomStringImpl;
return addSlowCase(WTFMove(string));
```
No extra ref but we're doing a use-after-move.

darinadler commented on Source/WebCore/dom/TreeScope.cpp:
```diff
@@ -441,7 +441,7 @@ RefPtr<Element> TreeScope::elementFromPoint(double clientX, double clientY, HitT
     }
 
-    return static_pointer_cast<Element>(WTFMove(node));
+    return uncheckedDowncast<Element>(WTFMove(node));
```
Why do we call some of these _unsafe_ and others _unchecked_? Don't both words mean the same thing?

geoffreygaren replied:
I often wonder the same thing!

cdumez replied:
We could call this unsafe, though this is not new to this PR. I think the main difference between the unsafe cast and the unchecked cast at the moment is that the unsafe one that zero validation while the unchecked one does validation in debug builds.

Also note that these unsafe casts are hopefully temporary as we are moving towards cast validation. unchecked casts have some use cases, after calling `is<>` like commented above. They are not ideal but sometimes the only way to write efficient code.

cdumez committed:
Rename static_pointer_cast() / static_reference_cast() to unsafeRefPtrDowncast() / unsafeRefDowncast()
https://bugs.webkit.org/show_bug.cgi?id=301991

Reviewed by Darin Adler.

Rename static_pointer_cast() / static_reference_cast() to unsafeRefPtrDowncast()
/ unsafeRefDowncast() to make it clear they are unsafe and reduce their use in
the codebase by leveraging:
1. downcast<>() which is truly safe and supports smart pointers nowadays
2. uncheckedDowncast<>() which is not ideal but better since it will at least validate
   the cast in debug builds.
3. New upcast() functions which are safe since it is safe to use static_cast for upcasts.

* Source/JavaScriptCore/bytecode/MetadataTable.h:
(JSC::MetadataTable::unlinkedMetadata const):
* Source/JavaScriptCore/inspector/InspectorProtocolTypes.h:
(Inspector::Protocol::BindingTraits<JSON::ArrayOf<T>>::runtimeCast):
* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py:
* Source/JavaScriptCore/inspector/scripts/tests/expected/type-requiring-runtime-casts.json-result:
* Source/WTF/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WTF/wtf/Ref.h:
(WTF::upcast):
(WTF::unsafeRefDowncast):
(WTF::uncheckedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
(WTF::static_reference_cast): Deleted.
* Source/WTF/wtf/RefPtr.h:
(WTF::upcast):
(WTF::unsafeRefPtrDowncast):
(WTF::uncheckedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
(WTF::static_pointer_cast): Deleted.
* Source/WTF/wtf/WeakPtr.h:
(WTF::weak_ptr_impl_cast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/WeakRef.h:
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/text/AtomString.h:
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::AtomStringImpl::addSlowCase):
* Source/WTF/wtf/text/AtomStringImpl.h:
(WTF::AtomStringImpl::add):
* Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebCore/bindings/js/JSDOMWrapperCache.h:
(WebCore::createWrapper):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::convertToExceptionOrNumericValue):
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementFromPoint):
(WebCore::TreeScope::elementsFromPoint):
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::splitAncestorsWithUnicodeBidi):
* Source/WebCore/editing/DeleteSelectionCommand.cpp:
(WebCore::firstInSpecialElement):
(WebCore::lastInSpecialElement):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::setComposition):
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplaceSelectionCommand::handleStyleSpans):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::lastTypingCommandIfStillOpenForTyping):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::finalizeFrame):
(WebCore::InspectorCanvas::markCurrentFrameIncomplete):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateDragAndDrop):
* Source/WebCore/page/ImageOverlayController.cpp:
(WebCore::ImageOverlayController::selectionQuadsDidChange):
* Source/WebCore/page/scrolling/ThreadedScrollingCoordinator.cpp:
(WebCore::ThreadedScrollingCoordinator::pageDestroyed):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::AudioFileReader::demuxWebMData const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
* Source/WebCore/rendering/RenderPtr.h:
(WTF::downcast):
(WebCore::static_pointer_cast): Deleted.
* Source/WebCore/rendering/style/StyleCrossfadeImage.cpp:
(WebCore::StyleCrossfadeImage::computedStyleValue const):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::computedStyleValue const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):
* Source/WebCore/svg/properties/SVGPrimitivePropertyAnimator.h:
(WebCore::SVGPrimitivePropertyAnimator::SVGPrimitivePropertyAnimator):
* Source/WebCore/svg/properties/SVGValuePropertyAnimator.h:
(WebCore::SVGValuePropertyAnimator::SVGValuePropertyAnimator):
* Source/WebCore/svg/properties/SVGValuePropertyListAnimator.h:
(WebCore::SVGValuePropertyListAnimator::SVGValuePropertyListAnimator):
* Source/WebKit/Shared/API/APIArray.h:
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
(webkitFeatureCreate):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
(webkitWebExtensionMatchPatternCreate):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::experimentalFeatures):
(WebKit::WebPreferences::internalDebugFeatures):
* Tools/TestWebKitAPI/Tests/WTF/Ref.cpp:
(TestWebKitAPI::TEST(WTF_Ref, StaticReferenceCastFromConstReference)):
(TestWebKitAPI::TEST(WTF_Ref, StaticReferenceCastFromRValueReference)):

Canonical link: https://commits.webkit.org/302695@main
